### PR TITLE
Update per_v_transform_reduce_incoming|outgoing_e to take a reduction operator.

### PR DESCRIPTION
**Pull request by seunghwak**
- Update `per_v_transform_reduce_incoming|outgoing_e` to take a reduction operator; currently reduction operators with a matching (element-wise) NCCL reduction operation are supported (i.e. SUM, MIN, and MAX).
- Deleted `evaluate_edge_op` and `evaluate_intersection_op` wrappers; these are initially invented to hide the difference in calling e_op with & without edge weights for weighted/unweighted graphs. Now edge weight is not part of a graph object, and every e_op takes a edge property value (which can be `thrust::nullopt` if no valid edge property). This indirection is no longer necessary.
- Rename `atomic_accumulate_edge_op_result` to `atomic_add_edge_op_result`

Non-breaking as primitives are currently not part of public API.

### CODE REVIEW

**Comment by seunghwak** [comment]
@naimnv This should support max reduction necessary for Leiden implementation.

**Comment by seunghwak** [comment]
rerun tests

**Review by ChuckHastings — APPROVED** [review]

File: cpp/tests/prims/mg_per_v_transform_reduce_incoming_outgoing_e.cu [hunk 1]
```diff
@@ -174,67 +191,124 @@ class Tests_MGPerVTransformReduceIncomingOutgoingE
...
 
-    if (cugraph::test::g_perf) {
-      RAFT_CUDA_TRY(cudaDeviceSynchronize());  // for consistent performance measurement
-      handle_->get_comms().barrier();
-      hr_timer.start("MG per_v_transform_reduce_outgoing_e");
-    }
+      if (cugraph::test::g_perf) {
+        RAFT_CUDA_TRY(cudaDeviceSynchronize());  // for consistent performance measurement
+        handle_->get_comms().barrier();
+        handle_->get_comms().barrier();
```
Comment (naimnv) [inline cpp/tests/prims/mg_per_v_transform_reduce_incoming_outgoing_e.cu:265]: Isn't second barrier redundant?

**Review by naimnv — APPROVED** [review]

File: cpp/tests/prims/mg_per_v_transform_reduce_incoming_outgoing_e.cu (hunk 1 above)
Comment (seunghwak) [reply cpp/tests/prims/mg_per_v_transform_reduce_incoming_outgoing_e.cu]: Yes and fixed, I guess this happened while resolving merge conflicts.

**Comment by ChuckHastings** [comment]
@gpucibot merge

**Pull request merged**